### PR TITLE
[Spark] add auto-compact post commit hook

### PR DESCRIPTION
**Pull request by nicklan**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
Adds code to support Auto Compaction.

Auto compaction combines small files within partitions to reduce problems due to a proliferation of small files. Auto compaction is implemented as a post commit hook, and so occurs after the write to a table has succeeded. It runs synchronously on the cluster that has performed the write. 

You can control the output file size by setting the  spark.databricks.delta.autoCompact.maxFileSize.

Auto compaction is only triggered for partitions or tables that have at least a certain number of small files. You can optionally change the minimum number of files required to trigger auto compaction by setting spark.databricks.delta.autoCompact.minNumFiles.

This PR creates a post commit hook, which runs an `OptimizeExecutor` (from `OptimizeTableCommand.scala`), which will do the compaction.

### Details
We add a post-commit hook in TransactionalWrite, that will check if auto-compaction is needed. If the configs are set such that the write meets the criteria (i.e. AC is enabled, enough small files exist, etc) then partitions that meet the criteria will be reserved, and used to make an OptimizeExecutor targeting those partitions, and with the appropriate config values.

This runs and will compact the files. Partitions are then released for future compactions to consider.

AutoCompact is *disabled* by default

### Configs
There are a number of new configs introduced by this PR, all with prefix spark.databricks.delta.autoCompact.  Through a lot of experimentation and user feedback, we found these values to work well across a large range of tables and configurations.

- `autoCompact.enabled`: should auto compaction run? (default: false)
- `autoCompact.maxFileSize`: Target file size produced by auto compaction (default: 128 MB)
- `autoCompact.minFileSize`: Files which are smaller than this threshold (in bytes) will be grouped together and rewritten as larger files by the Auto Compaction. (default: half of maxFileSize)

## How was this patch tested?
Unit tests in AutoCompactionSuite

## Does this PR introduce _any_ user-facing changes?

Yes, please see the [Description](#description)

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+  /**
+   * This method returns the type of Auto Compaction to use on a delta table or returns None
+   * if Auto Compaction is disabled.
+   * Prioritization:
+   *   1. The highest priority is given to [[DeltaSQLConf.DELTA_AUTO_COMPACT_ENABLED]] config.
+   *   2. Then we check if the deprecated property `DeltaConfigs.AUTO_OPTIMIZE` is set. If yes, then
+   *      we return [[AutoCompactType.Legacy]] type.
+   *   3. Then we check the table property [[DeltaConfigs.AUTO_COMPACT]].
+   *   4. If none of 1/2/3 are set explicitly, then we return None
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:55]: It never returns None

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+   */
+  def getAutoCompactType(conf: SQLConf, metadata: Metadata): Option[AutoCompactType] = {
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:57]: Option[AutoCompactType] -> AutoCompactType

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  private[delta] val OP_TYPE = "delta.commit.hooks.autoOptimize"
+
+  /**
+   * This method returns the type of Auto Compaction to use on a delta table or returns None
+   * if Auto Compaction is disabled.
+   * Prioritization:
+   *   1. The highest priority is given to [[DeltaSQLConf.DELTA_AUTO_COMPACT_ENABLED]] config.
+   *   2. Then we check if the deprecated property `DeltaConfigs.AUTO_OPTIMIZE` is set. If yes, then
+   *      we return [[AutoCompactType.Legacy]] type.
+   *   3. Then we check the table property [[DeltaConfigs.AUTO_COMPACT]].
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:54]: The new flag shouldn't have priority?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala [hunk 4]
```diff
@@ -388,6 +389,21 @@ trait DeltaConfigsBase extends DeltaLogging {
     _ > 0,
     "needs to be a positive integer.")
 
+  /**
+   * Enable auto compaction for a Delta table. When enabled, we will check if files already
+   * written to a Delta table can leverage compaction after a commit. If so, we run a post-commit
+   * hook to compact the files.
+   *  It can be enabled by setting the property to `true`
+   * Note that the behavior from table property can be overridden by the config:
+   * [[org.apache.spark.sql.delta.sources.DeltaSQLConfEdge.DELTA_AUTO_COMPACT_ENABLED]]
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaConfig.scala:398]: `org.apache.spark.sql.delta.sources.DeltaSQLConfEdge` I guess this is from your private code

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+  private[delta] val OP_TYPE = "delta.commit.hooks.autoOptimize"
+
+  /**
+   * This method returns the type of Auto Compaction to use on a delta table or returns None
+   * if Auto Compaction is disabled.
+   * Prioritization:
+   *   1. The highest priority is given to [[DeltaSQLConf.DELTA_AUTO_COMPACT_ENABLED]] config.
+   *   2. Then we check if the deprecated property `DeltaConfigs.AUTO_OPTIMIZE` is set. If yes, then
+   *      we return [[AutoCompactType.Legacy]] type.
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:53]: I also prefer `Legacy`, but it is currently named `Enabled`

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+  private[hooks] val DISABLED = "false"
+
+  /**
+   * This represents the legacy auto-compact behaviour:
+   * 1. MAX_FILE_SIZE is configurable and defaults to 128 MB unless overridden.
+   * 2. MIN_FILE_SIZE is configurable and defaults to MAX_FILE_SIZE / 2 unless overridden.
+   * Note: User can use DELTA_AUTO_COMPACT_MAX_FILE_SIZE to override this value.
+   */
+  case object Enabled extends AutoCompactType {
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:234]: Should we have two distinct objects? Enabled and Legacy to make it clear?

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+ */
+sealed trait AutoCompactType {
+  val configValueStrings: Seq[String]
+  /**
+   * This is the default maximum file size for Auto Compaction.
+   * 1. MAX_FILE_SIZE is configurable and defaults to 16 MB unless overridden.
+   * 2. MIN_FILE_SIZE is configurable and defaults to MAX_FILE_SIZE / 2 unless overridden.
+   * Note: User can use DELTA_AUTO_COMPACT_MAX_FILE_SIZE to override this value.
+   */
+  val defaultMaxFileSize: Long = 16 * 1024 * 1024
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:221]: I don't see any usage of defaultMaxFileSize

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    // If user-facing conf is set to something, use that value.
+    val autoCompactTypeFromConf =
+      conf.getConf(DeltaSQLConf.DELTA_AUTO_COMPACT_ENABLED).map(AutoCompactType(_))
+    if (autoCompactTypeFromConf.nonEmpty) return autoCompactTypeFromConf.get
+
+    // If user-facing conf is not set, use what table property says.
+    val deprecatedFlag = DeltaConfigs.AUTO_OPTIMIZE.fromMetaData(metadata)
+    val autoCompactTypeFromPropertyOrDefaultValue = deprecatedFlag match {
+      case Some(true) =>
+        Some(AutoCompactType.Enabled)
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:67]: For both, auto_optimize and auto_compact we are returning `Enabled`.

**Comment by felipepessoto** [comment]
@nicklan, we have another Auto Compact PR: https://github.com/delta-io/delta/pull/1156
Do you know what are the differences? Is this implementation the one used by Databricks, meaning it has been tested extensively?

**Comment by nicklan** [comment]
@felipepessoto thanks for the review! Yes, this is based on the implementation used by Databricks and so has been battle tested in many production workloads.  We absolutely considered #1156, and are grateful for the contribution. However, due to our extensive experience with this implementation we felt this was the best path forward.

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 1 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:55]: We actually do. If you look [here](https://github.com/delta-io/delta/pull/2414/files#diff-1506ccd714736b1ce36574203bccc6cc89c66d80b0bdec3b1eb5da6c653e06ccR248) in the `apply` method for `AutoCompactType`, when we pass it `Disabled` it returns `None`

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 2 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:57]: See comment about returning None

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 6 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:234]: Legacy is a, well, legacy name, there are only `Enabled`, or `Disabled`

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 9]
```diff
@@ -471,6 +471,8 @@ object DeltaOperations {
 
   sealed abstract class OptimizeOrReorg(override val name: String, predicates: Seq[Expression])
     extends OperationWithPredicates(name, predicates)
+  /** parameter key to indicate whether it's an Auto Compaction */
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:474]: to indicate Auto Compaction

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala [hunk 10]
```diff
@@ -482,10 +484,12 @@ object DeltaOperations {
   /** Recorded when optimizing the table. */
   case class Optimize(
       predicate: Seq[Expression],
-      zOrderBy: Seq[String] = Seq.empty
+      zOrderBy: Seq[String] = Seq.empty,
+      auto: Boolean = false
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:488]: Think `autoCompact` would be more meaningful

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 11]
```diff
@@ -350,6 +350,11 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     checkDeletionVectorFilesHaveWideBounds = false
   }
 
+  /** The set of distinct partitions that contain added files by current transaction. */
+  protected[delta] var partitionsAddedToOpt: Option[mutable.HashSet[Map[String, String]]] = None
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:354]: Remove `Opt` suffix (it's Scala - a type-heavy language after all 😉 )

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 12]
```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
     }
   }
 
+  def reportAutoCompactStatsError(e: Throwable): Unit = {
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:868]: Gotcha! That proves my point about calling the param `autoCompact` 😉

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 13]
```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
...
+    recordDeltaEvent(deltaLog, "delta.collectStats", data = Map("message" -> e.getMessage))
+    logError(e.getMessage)
+  }
+
+  def collectAutoOptimizeStats(numAdd: Long, numRemove: Long, actions: Iterator[Action]): Unit = {
+    // Early exit if no files were added or removed.
+    if (numAdd == 0 && numRemove == 0) return
+    val collector = createAutoCompactStatsCollector()
+    if (collector.isInstanceOf[DisabledAutoCompactPartitionStatsCollector]) return
+    AutoCompactPartitionStats.instance(spark)
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:878]: Is `instance` needed?! Why not use `apply`?

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 14]
```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
...
+      .collectPartitionStats(collector, deltaLog.tableId, actions)
+  }
+
+  /**
+   * A subclass of AutoCompactPartitionStatsCollector that's to be used if the config to collect
+   * auto compaction stats is turned off. This subclass intentionally does nothing.
+   */
+  class DisabledAutoCompactPartitionStatsCollector extends AutoCompactPartitionStatsCollector{
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:886]: A whitespace before closing `{`

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 15]
```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
...
+    override def collectPartitionStatsForAdd(file: AddFile): Unit = {}
+    override def collectPartitionStatsForRemove(file: RemoveFile): Unit = {}
+    override def finalizeStats(tableId: String): Unit = {}
+  }
+
+  def createAutoCompactStatsCollector(): AutoCompactPartitionStatsCollector = {
+    try {
+      if (spark.conf.get(DeltaSQLConf.DELTA_AUTO_COMPACT_RECORD_PARTITION_STATS_ENABLED)) {
+        val minFileSize = spark.conf.get(DeltaSQLConf.DELTA_AUTO_COMPACT_MIN_FILE_SIZE)
+              .getOrElse(Long.MaxValue)
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:896]: Why not the default value?

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala [hunk 16]
```diff
@@ -860,6 +865,46 @@ trait OptimisticTransactionImpl extends TransactionalWrite
...
+        return AutoCompactPartitionStats.instance(spark)
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:897]: `apply` 🙏

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 17]
```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:2]: 2024?

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 8 above)
Comment (felipepessoto) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:67]: If we are not splitting Enabled in two (Enabled, Legacy), this comment is not valid anymore

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+import org.apache.spark.sql.delta.stats.AutoCompactPartitionStats
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.catalog.CatalogTable
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A trait for post commit hook which compacts files in a Delta table. This hook acts as a cheaper
+ * version of the OPTIMIZE command, by attempting to compact files to a smaller size.
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:35]: a smaller size? Not bigger?

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+ *
+ * Auto Compact chooses files to compact greedily by looking at partition directories which
+ * have the largest number of files that are under a certain size threshold and launches a bounded
+ * number of optimize tasks based on the capacity of the cluster.
+ */
+trait AutoCompactBase extends PostCommitHook with DeltaLogging {
+
+  override val name: String = "Auto Compact"
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:43]: Compaction?

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+            AutoCompactType(v)
+          case None =>
+            AutoCompactType(AutoCompactType.DISABLED) // Default to disabled
+        }
+    }
+    autoCompactTypeFromPropertyOrDefaultValue
+  }
+
+  private[hooks] def shouldSkipAutoCompact(
+      autoCompactTypeOpt: Option[AutoCompactType],
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:84]: Remove `Opt` suffix

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala [hunk 21]
```diff
@@ -0,0 +1,376 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:2]: 2024?

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+ *   2. If all 'maxNumPartitions' are occupied, the partition stats of least recently used tables
+ *      will be evicted until the used partitions fall back below to 'maxNumPartitions'.
+ *   3. The un-partitioned tables are treated as tables with single partition.
+ * @param maxNumTablePartitions The hash space of partition key to reduce memory usage per table.
+ * @param maxNumPartitions The maximum number of partition that can be occupied.
+ */
+class AutoCompactPartitionStats(
+    private var maxNumTablePartitions: Int,
+    private var maxNumPartitions: Int
+) {
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:55]: Why on a separate line while the other class `PartitionStat` has them to close the input arguments?! Be consistent 🙏

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala [hunk 23]
```diff
@@ -0,0 +1,376 @@
...
+      var numFiles: Long,
+      var wasAutoCompacted: Boolean = false) {
+
+    /**
+     * Determine whether this partition can be autocompacted based on the number of small files or
+     * if this [[AutoCompactPartitionStats]] instance has not auto compacted it yet.
+     * @param minNumFiles The minimum number of files this table-partition should have to trigger
+     *                    Auto Compaction in case it has already been compacted once.
+     */
+    def hasSufficientSmallFilesOrHasNotBeenCompacted(minNumFiles: Long): Boolean =
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:77]: `...OrNotCompactedYet`?

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala [hunk 24]
```diff
@@ -0,0 +1,376 @@
...
+   * configuration.
+   */
+  private[delta] def resetTestOnly(): Unit = resetTestOnly(maxNumTablePartitions, maxNumPartitions)
+}
+
+object AutoCompactPartitionStats {
+  private var _instance: AutoCompactPartitionStats = null
+
+  /** The thread safe constructor of singleton. */
+  def instance(spark: SparkSession): AutoCompactPartitionStats = {
```
Comment (jaceklaskowski) [inline spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:363]: Why synchronized? Why is there a need for a singleton?

File: spark/src/test/scala/org/apache/spark/sql/delta/AutoCompactSuite.scala [hunk 25]
```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
```
Comment (jaceklaskowski) [inline spark/src/test/scala/org/apache/spark/sql/delta/AutoCompactSuite.scala:2]: 2024?

**Comment by felipepessoto** [comment]
This is not in the roadmap. Is it planned to 3.1 release?

File: spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 26]
```diff
@@ -196,6 +196,130 @@ trait DeltaSQLConfBase {
...
+  val DELTA_AUTO_COMPACT_MODIFIED_PARTITIONS_ONLY_ENABLED =
+    buildConf("autoCompact.modifiedPartitionsOnly.enabled")
+      .internal()
+      .doc(
+        s"""When enabled, Auto Compaction only works on the modified partitions of the delta
+           |transaction that triggers compaction.""".stripMargin)
+      .booleanConf
+      .createWithDefault(true)
+
+  val DELTA_AUTO_COMPACT_NON_BLIND_APPEND_ENABLED =
```
Comment (vkorukanti) [inline spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:294]: test with this property enabled?

File: spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 27]
```diff
@@ -196,6 +196,130 @@ trait DeltaSQLConfBase {
...
+    buildConf("autoCompact.nonBlindAppend.enabled")
+      .internal()
+      .doc(
+        s"""When enabled, Auto Compaction is only triggered by non-blind-append write
+           |transaction.""".stripMargin)
+      .booleanConf
+      .createWithDefault(false)
+
+  val DELTA_AUTO_COMPACT_MAX_NUM_MODIFIED_PARTITIONS =
```
Comment (vkorukanti) [inline spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:303]: same test comment here.

**Review by vkorukanti — APPROVED** [review]
lgtm after addressing all pending review comments.

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 11 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:354]: We use this fairly consistently:
- https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/catalog/DeltaTableV2.scala#L62
- https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala#L63
- https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/MetadataCleanup.scala#L63

We could propose something here, more generally, but for now I'll leave it as is

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 20 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:84]: see above

File: spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala (hunk 26 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:294]: Have a follow-up for some other testing, which we will do post merge

File: spark/src/test/scala/org/apache/spark/sql/delta/AutoCompactSuite.scala (hunk 25 above)
Comment (tdas) [reply spark/src/test/scala/org/apache/spark/sql/delta/AutoCompactSuite.scala:2]: aah... this is a global problem for all files. let fix this in a separate PR for all files.

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala (hunk 24 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:362]: Multiple threads could be committing simultaneously, but we want to keep all the stats together. This gives us a safe way to have a single source of stats across all transactions

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 13 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:878]: see https://github.com/delta-io/delta/pull/2414/files#r1446659873

File: spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala (hunk 16 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala:898]: see https://github.com/delta-io/delta/pull/2414/files#r1446659873

File: spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala (hunk 19 above)
Comment (tdas) [reply spark/src/main/scala/org/apache/spark/sql/delta/hooks/AutoCompact.scala:44]: I prefer "Auto Compact".. consistent with hook and feature name... not to be confused with optimize / compaction stuff which runs offline.

**Comment by tdas** [comment]
Overall, this PR looks pretty good to me other than small naming improvements that others have pointed out (@felipepessoto @jaceklaskowski thank you so so much for your reviews). We are nearing the target [3.1](https://github.com/delta-io/delta/milestone/25) release date of Jan 24, and I would like to cut a release soon with this feature. Unless there is major objections to the core logic of the PR, can we merge this?

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 9 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:478]: re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge

File: spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala (hunk 10 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/DeltaOperations.scala:492]: re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala (hunk 22 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:55]: re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge

File: spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala (hunk 23 above)
Comment (nicklan) [reply spark/src/main/scala/org/apache/spark/sql/delta/stats/AutoCompactPartitionStats.scala:77]: re https://github.com/delta-io/delta/pull/2414#issuecomment-1883922342, perhaps we can address post merge

**Comment by felipepessoto** [comment]
I only have one concern, my comment:

> The new flag shouldn't have priority?

It that right? Is it the same behavior of Databricks? So that could be a reason to keep compatibility.

I don't see any previous usage of that flag, for example for optimized write we ignored the old flag and don't use it at all.

**Comment by felipepessoto** [comment]
And a nit, the AUTO_COMPACT could be placed close to the other autoOptimize settings

**Comment by nicklan** [comment]
> It that right? Is it the same behavior of Databricks? So that could be a reason to keep compatibility.

Yeah, this is to maintain compatibility with Databricks, so I think we should keep it as is

**Comment by nicklan** [comment]
> And a nit, the AUTO_COMPACT could be placed close to the other autoOptimize settings

Also makes sense, but is trickier than you might expect. We can absolutely look at cleaning this up post merge if possible

File: spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 28]
```diff
@@ -196,6 +196,130 @@ trait DeltaSQLConfBase {
...
+      .intConf
+      .checkValue(_ > 0, "The value of partitionStatsSize should be positive.")
+      .createWithDefault(64 * 1024)
+
+  val DELTA_AUTO_COMPACT_MAX_FILE_SIZE =
+    buildConf("autoCompact.maxFileSize")
+      .internal()
+      .doc(s"Target file size produced by auto compaction. The default value of this config" +
+        " is 128 MB.")
+      .longConf
```
Comment (felipepessoto) [inline spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:263]: Should we use bytesConf to be consistent with the optimized write?

Comment (felipepessoto) [reply spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:263]: I'd say to do it after 3.1 cut, but before final release, if possible

**Pull request closed without merging**